### PR TITLE
Query manager

### PR DESCRIPTION
**Pull request by nupurgunwant**
Hi @julie-sullivan @rachellyne ,
This pull request adds a file `query_manager.py` ( other than the registry features )
This file contains the following functions:
`get_all_query_names()` to view the saved queries from a user account ,
 `get_query()` to fetch information about a saved query,
 `delete_query()` to delete a saved query from user account
`post_query()` to post a query to user account

Steps to test:

- git clone https://github.com/nupurgunwant/intermine-ws-python.git
- cd intermine-ws-python
- python setup.py install
- pip install -r requirements.txt
- git checkout query_manager
- python(this will open the interactive mode of python, make sure the version is >=3)
==>from intermine import query_manager as qm
'Enter the account's API token and mine to which account is linked'
==>qm.get_all_query_names()
'returns the names of saved queries'
==>qm.get_query()
'enter name of query you want details about'
'returns details'
==>qm.delete_query()
'enter name of query you want to delete'
==>qm.post_query()
'enter the xml of the query you want to save on the account'

You may check your account after each function to check if the results are correct. Looking forward to all the suggestions. :)
Here's a screenshot in reference to the above steps.
screen shot 2018-05-16 at 11 20 32 am

PS : Merging this PR will add registry features as well.

### CODE REVIEW

**Comment by julie-sullivan** [comment]
Looks great! This is going to be super useful! :) I have some notes:

* Most people do not use python in the console. Bioinformaticians write scripts so they can reproduce their workflow. I like the interactive features! I would like to do both, but scripts are the main use case. Is it possible to enable, e.g. `delete_query(queryName)`?
* Can you add some comments to each function? Use the other python classes for the correct syntax. Better to do documentation as we go to help me see what's going on. Also make as self documenting as possible, e.g. descriptive variable names.
* You mentioned something like checkstyle for Python. Do you think it's a good idea to enable that now, when we are getting started? Your code is easy to read but I do see formatting inconsistencies. If you don't want to implement checkstyle, use the formatting you see elsewhere, e.g. spaces in tests `(1 == 1)`
* Can we add a bit of error handling? I see you've done some! But I'd like to add more, e.g these edge cases:
  * incorrect XML
  * there are rules for naming queries, e.g. no special characters I think?. You'll want to enforce those.
  * duplicate query names, see `get_unused_list_name()` for how to handle
  * trying to delete a query that doesn't exist.
  * invalid token
  * valid token, but user has no saved queries

**Comment by julie-sullivan** [comment]
Also, minor notes:

* I see lots of UTF8 handling in the list manager. Something to look into!
* We'll want to also add the ability to manage tags for queries.

**Comment by julie-sullivan** [comment]
We should also add some tests.

* We do not want to test the InterMine API
* We only want to test the Python client code.
* We have "live" tests that Travis runs that do test the InterMine API. Ignore those.
* We used to run tests against live FlyMine but these tests would fail everytime I updated FlyMine lol. Instead we want to use dummy data that will always be the same.

Given the way your code is structured, I suggest this:

* Refactor the code to have one class/method handle the web service requests
* For tests, replace with a dummy class/method that returns dummy results for testing.

As long as the dummy results you are testing look exactly like what the API would return, the tests will be useful.

Does that make sense?

@justinccdev Do you have Python testing thoughts?

**Comment by justinccdev** [comment]
You can run the current tests with

```python3 setup.py test```

If I remember, all the tests are run twice, once against a live backend and another time against a dummy backend.  Unfortunately, the code is very convoluted and it's hard to figure out what it's doing.  To run tests against just a dummy backend you will probably have to unpick bits of this structure, which I expect will be a big and intricate task.  But writing tests has to be done and the current tests are not that extensive, as far as I recall.  Good luck!

Also, standard style for Python as you probably already know is [PEP 8](https://www.python.org/dev/peps/pep-0008/).  Please use this if you are going to enforce a consistent style.  Again, enforcing this on the existing code might require lots of editing.  You might want to only do it on any new code you write, unless you are sure the test suite will pick up any mistakes.

**Comment by julie-sullivan** [comment]
@justinccdev What is your opinion on checkstyle for Python?

File: intermine/query_manager.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    example:
+
+        >>>qm.post_query('<query name="" model="genomic" view="Gene.length Gene.symbol" \
+           longDescription="" sortOrder="Gene.length asc"></query>')
+    """
+    x = "http://registry.intermine.org/service/instances/" + mine
+    r = requests.get(x)
+    dict = yaml.load(r.text)
+    name = []
+    for i in range(13,len(xml)):
```
Comment (julie-sullivan) [inline intermine/query_manager.py]: What is 13? A magic number!

https://en.wikipedia.org/wiki/Magic_number_(programming)#Unnamed_numerical_constants

File: intermine/query_manager.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    '''
+    x = "http://registry.intermine.org/service/instances/" + mine
+    r = requests.get(x)
+    dict = yaml.load(r.text)
+    link = dict["instance"]["url"] + "/service/user/queries/" + name +\
+     "?token=" + token
+    requests.delete(link)
+    '''
+
+def xml_to_link(xml):
```
Comment (julie-sullivan) [inline intermine/query_manager.py]: To my non-python eyes, this looks fragile to me. Should we use a library instead?

https://stackoverflow.com/questions/5607551/how-to-urlencode-a-querystring-in-python

File: intermine/query_manager.py (hunk 1 above)
Comment (nupurgunwant) [reply intermine/query_manager.py]: No, actually in every xml string, the 13th element will be the one where name starts.
example: '<query name="xxx" model=.....
< = 0th position
q = 1st position
u = 2nd position
....
" = 12th position
x = ***13th position***

Comment (julie-sullivan) [reply intermine/query_manager.py]: Maybe this is clearer explanation as to why this is a bad thing:

https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad

File: intermine/query_manager.py (hunk 2 above)
Comment (nupurgunwant) [reply intermine/query_manager.py]: Any of these libraries is not fully satisfying our requirements. For instance 'urllib.request' isn't converting "/" to "%2F". 
Also, I was avoiding other dependencies( although urllib.request is inbuilt in python3 ) 
So, even if we find such a library it would just make things complicated.

Comment (julie-sullivan) [reply intermine/query_manager.py]: @justinccdev What are your thoughts? I would prefer to not have a lot of dependencies, but XML is pants.

File: intermine/query_manager.py (hunk 1 above)
Comment (nupurgunwant) [reply intermine/query_manager.py]: Oh right! This seems better, will do the necessary change

File: intermine/query_manager.py (hunk 2 above)
Comment (nupurgunwant) [reply intermine/query_manager.py]: Well, 'urllib.request' is doing the job right. (inspite of the fact that it doesn't convert "/" ).
So, its your choice whether we should use this inbuilt library or define a function in the code itself.

**Comment by justinccdev** [comment]
Some comments from a very brief look at the code

* Why are you using pyaml to load JSON?  This is unnecessary and exposes the user to unnecessary danger if the registry were ever [hacked](http://pyyaml.org/wiki/PyYAMLDocumentation).  Please use the built in json.load() instead.

* Use libraries whenever possible.  I know the previous code had no existing requirements but this was bad practice.

* All your if statements have unnecessary ()

* Don't write your own quoter (xml_to_link).  Use [urllib.quote_plus](https://stackoverflow.com/questions/5607551/how-to-urlencode-a-querystring-in-python/5607708) or similar, though this particular function will be a bit tricky when ensuring the code remains both python 2 and 3 compatible.

* If you're working with xml then parse it with lxml or similar.  Don't manipulate it with string functions, this is terrible for readability and makes everythign more fragile.

File: intermine/query_manager.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        raw = json.loads(r.text)
+        for key in raw['queries'].keys():
+            if key == name:
+                flag = 1
+        if flag == 0:
+            print("Incorrect xml (Note: name should contain no special symbol\
+             and should be defined first)")
+
+
+position = 13
```
Comment (julie-sullivan) [inline intermine/query_manager.py]: define this variable before it's used.

**Comment by julie-sullivan** [comment]
@justinccdev What does the `do not use bare except` mean and how do we fix this? Can we tell PEP8 to ignore because it seems dumb.

**Comment by justinccdev** [comment]
If we are talking about query_manager.py, the `do not use bare except` points out a genuine issue with the code, in that it appears to test if the token is valid by looking for a queries key in the returned XML but will fail for lots of other reasons as well, including if the mine is offline.  Error checking and handling different cases needs to be better.  And some other observations

* You are still using position to chop the xml.  Please don't do this.
* I see this chunk of code is just in query_manager directly.  How is it being invoked?
* There should be tests as part of this PR before it is accepted.

**Comment by justinccdev** [comment]
And also it doesn't look that code works anyway since you attempt to use a parameter (n) which is undefined!  I think would misleadingly give you an 'Invalid Token' error, which is the kind of problem that the warning is highlighting.

**Comment by nupurgunwant** [comment]
Hi @justinccdev, Thanks for the review 

- I have modified `except` to now return the type of error. Is that better or you prefer it by specifying different error message for different errors? 
- I used lxml, like you suggested for parsing the xml. Thanks for the suggestion, it seems much better now. However, I was facing merge conflict right now while adding the requirement of lxml in requirement.txt
- The chunk of that I have added directly in query_manager is invoked while importing the query_manager file. Do you prefer making a separate function for that?
- I will be adding the tests in the following week

Do let me know if I left out anything.

**Comment by nupurgunwant** [comment]
Update:
- mine and token are now taken in the following function
>>>from intermine import query_manager as qm
>>>qm.save_mine_and_token('flymine','k136n1HfFd31n6O4han1')
- I have added tests for query_manager, and refactored code to meet that(not making much difference in the output). 
The way I am testing an account is by creating a demo account(test is written such that it can be any account). This will be used for testing all the functions.
Is there any loophole in this method? I did read about mock class and monkey patching but that wasn't matching with current situation and might require some major refactoring.

Also:
- I will add the registry refactoring and testing in a separate branch(registry_updated) to avoid merge issues.

File: tests/test_query_manager.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
+import unittest
+
+from ..intermine import query_manager as qm
+qm.save_mine_and_token('flymine', 'k136n1HfFd31n6O4han1')
```
Comment (julie-sullivan) [inline tests/test_query_manager.py]: Unfortunately we can't use flymine as a test mine.

1. It's a live InterMine used by people. 
2. FlyMine's model might change
3. FlyMine might not live forever.

Instead, do the thing I suggested. Make a dummy mine in your code that has hardcoded valid tokens, returns good and bad results, etc.

This will require that you refactor your code.

**Comment by julie-sullivan** [comment]
@justinccdev Please chime in because I don't know the best practices.

@nupurgunwant We want to test your code only, not the InterMine API. The API is tested elsewhere. However, your methods (a) make the API request (b) process the request. This means that we can't test your code only, because the API code is wrapped up with it.

One option is to make a dummy mine. Instead of testing flymine.org, we could test localhost/biotestmine for example. Unfortunately this won't work because we don't want to have to prop up a webapp just to run python tests. There are "live" tests in the Python code now that do this, and are run in Travis.

Instead I would like you to refactor your code. This is per our discussion yesterday. 

Current status:

```
def get_all_query_names():
 // builds mine request URL
 // makes request
 // processes request
```
Change to be:

```
def get_all_query_names():
 // calls get_queries(mine)
 // parses the results -- in this case, just printing

def get_queries():
 // makes the request to the mine
 // returns the reponse
 // IF TEST RETURN A DUMMY RESPONSE
```

**Comment by julie-sullivan** [comment]
What do we want to test? That @nupurgunwant's code does what it's supposed to. 

```
get_all_query_names()
get_query(name)
delete_query(name):
post_query(xml):
save_mine_and_token(m, t)
```

 `get_all_query_names()` is just printing, for example. So what exactly are we going to be testing?

**Comment by nupurgunwant** [comment]
Absolutely, I understand that using a live mine, such as 'flymine' was a bad idea.
I have refactored my code in order to handle a mine if its called 'mock'. I have added this as a condition in the present function itself.
As for example, here's how get_all_query_names() looks like now:
```
def get_all_query_names():
    if mine == 'mock':
        // manually defined 'dict' object
    else:
        // code to access a 'dict' object from a live mine which needn't be tested
    
    # operations with the 'dict' object which will be tested

```

Does this seem better? I have written and tested it for all functions .

**Comment by julie-sullivan** [comment]
@nupurgunwant Yes that looks good thanks!

**Comment by nupurgunwant** [comment]
Instructions for checking all functions ( since a lot has changed ):

- git clone https://github.com/nupurgunwant/intermine-ws-python.git
- cd intermine-ws-python
- python setup.py install
- pip install -r requirements.txt
- git checkout query_manager
- python(this will open the interactive mode of python, make sure the version is >=3)
==> from intermine import query_manager as qm
==> qm.save_mine_and_token('humanmine','T1p6E3FdZ689o8o9c0W7')
==> qm.get_all_query_names()
nnn
xxx
==> qm.get_query('nnn')
Columns:
Gene.briefDescription
Gene.name
Gene.symbol
==> qm.delete_query('nnn')
'nnn is deleted'
==> qm.get_query('nnn')
'No such query available'
==> qm.delete_query('nnn')
'No such query available'
==> qm.post_query('<query name="nnn" model="genomic" view="Gene.primaryIdentifier Gene.proteins.primaryAccession Gene.symbol" longDescription="" sortOrder="Gene.primaryIdentifier asc"></query>')
'nnn is posted'
==> qm.get_all_query_names()
nnn
xxx

File: intermine/query_manager.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+                resp = raw_input("Please enter response again [y/n]")
+            if resp == 'y':
+                count = 0
+
+    if count == 0:
+        # mock raw for testing
+        if mine == 'mock':
+            raw = {'queries': {'query1': 1, 'query2': 2, 'query3': 3}}
+        else:
+            link = dict["instance"]["url"] + "/service/user/queries?xml=" + \
```
Comment (julie-sullivan) [inline intermine/query_manager.py]: this parameter `xml` needs to be changed because it now accepts JSON queries. See https://github.com/intermine/intermine/pull/1815

What can we do? 

Many mines will have the `xml` parameter a long time. But it doesn't make any sense to have a parameter called XML that accepts JSON.

Do we have any ideas for solutions?
